cmd/promcheck/check.go

```diff
@@ -160,16 +161,39 @@ func (app *promcheckApp) checkRulesFromRuleFiles() error {
 		return err
 	}
 
+	var eg errgroup.Group
 	checkResults := []promcheck.CheckResult{}
+	resultChan := make(chan promcheck.CheckResult, len(ruleGroupsToCheck))
+
 	for _, group := range ruleGroupsToCheck {
-		checked, err := app.check.CheckRuleGroup(group)
-		if err != nil {
-			level.Error(app.logger).Log("msg", "failed to check rule groups", "file", group.File, "err", err)
-			return err
+		group := group // https://golang.org/doc/faq#closures_and_goroutines
+		eg.Go(func() error {
+			checked, err := app.check.CheckRuleGroup(group)
+			if err != nil {
+				level.Error(app.logger).Log("msg", "failed to check rule groups", "file", group.File, "err", err)
+				return err
+			}
+			for _, res := range checked {
+				resultChan <- res
+			}
+			app.report.AddTotalCheckedGroups(1)
+			return nil
+		})
+	}
+
+	go func() {
+		if err := eg.Wait(); err != nil {
```

Comment:
I think usually you don't want `eg.Wait` to be in a separate goroutine. However, with the results chennel it might be a bit different. I'd need to close at it more closely.